### PR TITLE
feat: Save Brillig execution state in ACVM

### PR DESCRIPTION
**Pull request by ggiraldez**
# Description

This PR changes ACVM to retain the Brillig solver state while executing a circuit.

## Problem

This is in preparation to support stepping into Brillig opcodes when debugging Brillig blocks (part of #3015). It also brings an additional benefit: the Brillig VM does not need to be restarted from scratch after resolving a foreign call, which should result in a performance improvement when executing circuits.

## Summary

The `BrilligSolver` struct was modified to hold a Brillig VM instance and this is kept in `ACVM` inside an `Option<>` while solving a Brillig ACIR opcode.

In order to be able to push the results of foreign calls to the Brillig VM without having to rebuild it, a new method was added to push a `ForeignCallResult` to the results vector.

These are additive changes and should not break existing code.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

Some pending considerations:

- It should now be possible to remove the `foreign_call_results` parameter when initializing a Brillig VM, since new results can be inserted while running.
- It may also make sense to use a different data structure since strictly speaking the Brillig VM would only need the results from the last pending call.
- The ACIR opcodes in ACVM no longer need to be mutable or even owned by the ACVM structure. Mutability was needed to receive and hold the results from foreign calls, but those are now directly sent to the running Brillig VM.

# PR Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by TomAFrench** [comment]
> The ACIR opcodes in ACVM no longer need to be mutable or even owned by the ACVM structure. Mutability was needed to receive and hold the results from foreign calls, but those are now directly sent to the running Brillig VM.

We actually have this already implemented in https://github.com/noir-lang/noir/pull/2872 although we're not in a big rush to get it merged due to the resulting serialisation change.

File: acvm-repo/acvm/src/pwg/brillig.rs
Comment (TomAFrench) [inline acvm-repo/acvm/src/pwg/brillig.rs:37]: I think that this function is trying to do too many things which is adding complexity into `solve_opcode`. We should probably handle the case where we're skipping the opcode separately.

File: acvm-repo/acvm/src/pwg/mod.rs
Comment (TomAFrench) [inline acvm-repo/acvm/src/pwg/mod.rs:292]: For example, compare this with pseudocode

```rust
if BrilligSolver::should_skip(witness, brillig)? {
  BrilligSolver::zero_out_brillig_outputs(witness, brillig)
} else {
  let solver = self.brillig_solver.take().unwrap_or_else(|| {
      BrilligSolver::new(witness, brillig, self.backend, self.instruction_pointer)
  });
  match solver.solve()? {
      BrilligSolverStatus::ForeignCallWait(foreign_call) => {
          // Save solver state to resume execution after foreign call.
          self.brillig_solver.insert(solver);
          return self.wait_for_foreign_call(foreign_call);
      }
      BrilligSolverStatus::InProgress => {
          unreachable!("Brillig solver still in progress")
      }
      BrilligSolverStatus::Finished => {
          // Write execution outputs to witness
          solver.finalize(witness, brillig)
      }
  }
}
```

We'll need to turn this big `let resolution = match` statement into a helper function so that we can use `?` syntax, but it's much easier to see what this code is doing imo.

File: acvm-repo/acvm/src/pwg/brillig.rs
Comment (ggiraldez) [reply acvm-repo/acvm/src/pwg/brillig.rs:34]: Yes, you're right! Those are left overs from a previous iteration.

File: acvm-repo/acvm/src/pwg/brillig.rs
Comment (ggiraldez) [reply acvm-repo/acvm/src/pwg/brillig.rs:37]: Agreed. I put everything together in an attempt to capture more of the logic to deal with Brillig blocks, but I agree it hurts readability.

File: acvm-repo/acvm/src/pwg/mod.rs
Comment (ggiraldez) [reply acvm-repo/acvm/src/pwg/mod.rs:292]: I like this! But there is one problem for which I haven't found a good idiomatic solution: the early return when waiting for foreign calls, which cannot go in the helper. I'm still familiarizing myself with Rust and don't know of a good way to handle this, short of introducing yet another return enum type (which I'd like to avoid). Maybe there is something in the stdlib?

Comment (TomAFrench) [reply acvm-repo/acvm/src/pwg/mod.rs:292]: Ah yeah, I hadn't thought of that. Maybe for the short term we could just move the brillig arm of this match into a helper function and return a `Result<Option<ForeignCallWaitInfo>, OpcodeResolutionError>`. We can then perform the early return based on whether the option is `Some` or `None`.

```rust
// New match arm
match self.solve_brillig_opcode(brillig) {
    Ok(Some(foreign_call)) => {
        return self.wait_for_foreign_call(foreign_call);
     }
     result => result.map(|_| ()),
}

// helper function
fn solve_brillig_opcode(
    &mut self,
    brillig: &Brillig,
) -> Result<Option<ForeignCallWaitInfo>, OpcodeResolutionError> {
    let witness = &mut self.witness_map;
    if BrilligSolver::should_skip(witness, brillig)? {
        BrilligSolver::zero_out_brillig_outputs(witness, brillig).map(|_| None)
    } else {
        let solver = self.brillig_solver.take().unwrap_or_else(|| {
            BrilligSolver::new(witness, brillig, self.backend, self.instruction_pointer)
        });
        match solver.solve()? {
            BrilligSolverStatus::ForeignCallWait(foreign_call) => {
                // Save solver state to resume execution after foreign call.
               self.brillig_solver.insert(solver);
                Ok(Some(foreign_call))
            }
            BrilligSolverStatus::InProgress => {
                unreachable!("Brillig solver still in progress")
            }
            BrilligSolverStatus::Finished => {
                // Write execution outputs to witness
                solver.finalize(witness, brillig).map(|_| None)
            }
        }
    }
}
```

File: acvm-repo/acvm/src/pwg/brillig.rs
Comment (ggiraldez) [reply acvm-repo/acvm/src/pwg/brillig.rs]: I have split `build_or_skip` as per your suggestion. Thanks!

File: acvm-repo/acvm/src/pwg/mod.rs
Comment (ggiraldez) [reply acvm-repo/acvm/src/pwg/mod.rs]: I steered the code in the direction you proposed, but I found a couple other problems:
- It's not possible to use `unwrap_or_else` because `BrilligSolver::new` may return an error when reading the inputs.
- I found no clean way to send `brillig` as an argument, because that's a reference inside `self`, which needs to be mutably borrowed for the call itself.

I'm not 100% happy with the resulting code, but I think it's better than before. In any case, it will probably mutate again when we add support for step-debugging through Brillig blocks.

File: acvm-repo/acvm/src/pwg/mod.rs
Comment (TomAFrench) [inline acvm-repo/acvm/src/pwg/mod.rs:318]:
```suggestion
                    self.brillig_solver = Some(solver);
```

This would be a more idiomatic way to do this.

File: acvm-repo/acvm/src/pwg/mod.rs
Comment (TomAFrench) [inline acvm-repo/acvm/src/pwg/mod.rs:314]:
```suggestion
            // If we're resuming execution after resolving a foreign call then
            // there will be a cached `BrilligSolver` to avoid recomputation.
            let mut solver: BrilligSolver<'_, B> = match self.brillig_solver.take() {
                Some(solver) => solver,
                None => {
                    BrilligSolver::new(witness, brillig, self.backend, self.instruction_pointer)?
                }
            };
```

Let's add a note on why we sometimes have this solver. It's a matter of taste but it would be nice to have the `Some` branch first as well.

File: acvm-repo/acvm/src/pwg/mod.rs
Comment (TomAFrench) [inline acvm-repo/acvm/src/pwg/mod.rs:318]:
```suggestion
                    // Cache the current state of the solver
```

File: acvm-repo/brillig_vm/src/lib.rs
Comment (TomAFrench) [inline acvm-repo/brillig_vm/src/lib.rs:140]: I think there's an argument to be made to remove the `foreign_call_results` argument from `VM::new` once we have this function.

**Comment by TomAFrench** [comment]
Ah, the suggestions are a bit broken as I made them in a codespace. Will fix.

File: acvm-repo/acvm/src/pwg/mod.rs
Comment (TomAFrench) [inline acvm-repo/acvm/src/pwg/mod.rs:317]:
```suggestion
                    // Cache the current state of the solver
                     self.brillig_solver = Some(solver);
```

File: acvm-repo/acvm/src/pwg/mod.rs
Comment (TomAFrench) [inline acvm-repo/acvm/src/pwg/mod.rs:314]:
```suggestion
            // If we're resuming execution after resolving a foreign call then
            // there will be a cached `BrilligSolver` to avoid recomputation.
            let mut solver: BrilligSolver<'_, B> = match self.brillig_solver.take() {
                Some(solver) => solver,
                None => {
                    BrilligSolver::new(witness, brillig, self.backend, self.instruction_pointer)?
                }
            };
```

File: acvm-repo/acvm/src/pwg/mod.rs
Comment (TomAFrench) [reply acvm-repo/acvm/src/pwg/mod.rs]: Yeah, thanks for making these changes. I'm happy with the current state of this code, it's not perfect but given the constraints I think it's as readable as we can get for now. Let's leave it as is, as you say, we're going to be modifying it again in future.

> I found no clean way to send brillig as an argument, because that's a reference inside self, which needs to be mutably borrowed for the call itself.

Yeah, it's a little bit of a pain. If we passed in references to individual fields from `self` we could get around this but  that's just another tradeoff.

**Comment by TomAFrench** [comment]
This is looking good, I'm happy to merge once the suggestions are applied. We've got a slight issue with having two methods of feeding in foreign call results into the VM which we may want to address but that can also go in a separate PR (as a cleanup after #2872)

File: acvm-repo/brillig_vm/src/lib.rs
Comment (ggiraldez) [reply acvm-repo/brillig_vm/src/lib.rs:140]: Agreed :+1: We should remove the check in that case, to allow feeding future results before the call opcodes are executed.

File: acvm-repo/acvm/src/pwg/mod.rs
Comment (ggiraldez) [reply acvm-repo/acvm/src/pwg/mod.rs:317]: Oh, yes, thank you!

**Comment by TomAFrench** [comment]
@ggiraldez Could you mark future PRs as "Allow edits from maintainers" please? I'd like to merge this but am blocked by the lifetime error, ideally I'd be able to fix this up myself.

**Comment by TomAFrench** [comment]
You'll need to merge master into your branch as the error is cause by a conflict with #3030

**Comment by ggiraldez** [comment]
> @ggiraldez Could you mark future PRs as "Allow edits from maintainers" please? I'd like to merge this but am blocked by the lifetime error, ideally I'd be able to fix this up myself.

We made the fork in our organization @manastech, and apparently that option is not available for organization forks :(

**Comment by TomAFrench** [comment]
Ah ok, no worries then. :)

**Comment by ggiraldez** [comment]
> You'll need to merge master into your branch as the error is cause by a conflict with #3030

Ok, I'm having trouble with lifetimes now and I'm not being able to solve the issue. The `BrilligSolver` needs to hold a (transitive) reference to the Brillig opcodes inside the `ACVM` and that holds a mutable borrow when calling `solve_brillig_opcode()` thus not allowing any other borrows afterwards. Intuitively I know this should work, but I also understand why the compiler complains. The same thing happens in `solve()` when calling `solve_opcode()` btw.

I'm lost on how to fix this issue without resorting to copying the Brillig opcodes to spawn the VM (if that's even possible). Maybe using a `RefCell` to hold the opcodes in the ACVM?

**Comment by TomAFrench** [comment]
Oof, yeah that's unfortunate. I'll have a think about how to address this.

**Comment by TomAFrench** [comment]
Good news, I had a suspicion that https://github.com/noir-lang/noir/pull/2872 would fix this issue for us and it has. Merging master into this branch should be pretty easy now.

**Comment by ggiraldez** [comment]
> Good news, I had a suspicion that #2872 would fix this issue for us and it has. Merging master into this branch should be pretty easy now.

Whew! Ok, the merge was pretty painless.  Some integration tests are failing now though (`hello_world` and `codegen-verifier` AFAICS) but I don't see how that's related to the changes in this PR.

I think now it would be pretty safe to remove the `foreign_call_results` in `BrilligVM` as you suggested above. For the solver in fact, that's passed as `vec![]`.

**Review by TomAFrench — APPROVED** [review]
LGTM. We can address any issues related to foreign calls in a followup PR

**Pull request merged**
